File: 0859-Buddy Strings/0859-Buddy Strings.cpp
class Solution {
public:
    bool buddyStrings(string A, string B) {
        if (A.size() != B.size()) {
            return false;
        }
        vector<int> counts(26);
        bool duplicate = false;
        int diff = 0, first = -1, second = -1;
        for (int i = 0; i < A.size(); ++i) {
            if (A[i] != B[i]) {
                ++diff;
                if (diff == 1) {
                    first = i;
                }
                else if (diff == 2) {
                    second = i;
                }
                else {
                    return false;
                }
            }
            else if (++counts[A[i] - 'a'] > 1){
                duplicate = true;
            }
        }
        return (diff == 0 && duplicate == true) || (diff == 2 && A[first] == B[second] && A[second] == B[first]);
    }
};

File: 0801-0900/0859-Buddy Strings/0859-Buddy Strings.py
class Solution:
    def buddyStrings(self, A: str, B: str) -> bool:
        if len(A) != len(B):
            return False
        visited = set()
        duplicate = False
        diff = first = second = 0
        for i in range(len(A)):
            if A[i] != B[i]:
                diff += 1
                if diff == 1:
                    first = i
                elif diff == 2:
                    second = i
                else:
                    return False
            else:
                if A[i] in visited:
                    duplicate = True
                else:
                    visited.add(A[i])
        return (diff == 0 and duplicate == True) or (diff == 2 and A[first] == B[second] and A[second] == B[first])
